client.go

```diff
@@ -205,11 +206,35 @@ func checkErrorResponse(response []byte) error {
 	return nil
 }
 
+// request fetches an access token from the cache or re-authenticate when the access token is not available in the cache
+// and sends a request using that token
+func (c Client) request(ctx context.Context, method string, url string, userAgent string, params map[string]string, bodyStr string) ([]byte, error) {
+	accessToken := getCachedAccessToken(c.Username, c.ApiEndpoint)
+	var response []byte
+	var err error
+	var responseCode int
+	if len(accessToken) > 0 {
+		response, err, responseCode = request(ctx, accessToken, method, url, userAgent, params, bodyStr)
+	}
+	if len(accessToken) == 0 || responseCode == http.StatusUnauthorized {
+		deleteAccessTokenFromCache(c.Username, c.ApiEndpoint)
+		// Refreshing the access token as it is expired
```

Comment:
Can we check for expired token before making a request? Since we already have an expiration time cached it would be more efficient than doing a round trip to the server.

Reply:
getCachedAccessToken(c.Username, c.ApiEndpoint) returns tokens that are not expired based on their expiration time (ttlcache gets rid of the tokens that are expired automatically) . We could face a  'http.StatusUnauthorized' if the token is 'rejected' by the server although based on their expiration time they should still be valid(which happened once in the past).

Reply:
I see. Looking through ttlcache readme I think we're supposed to call `cache.Start()` or `cache.DeleteExpired()` to evict entries? Or do you mean `getCachedAccessToken` would not return invalid entry and we'll just overwrite it with a new token, instead of deleting it?

Reply:
cache.Start() and cache.DeleteExpired() are not needed when we set the ttl during the insertion of the token in the cache. (This is tested by https://github.com/firebolt-db/firebolt-go-sdk/blob/6749d13ec592b525eca60ee855a180e54c71bad8/client_test.go#L83-L116 ).